### PR TITLE
fix: Fix fp8 memory fragmentation

### LINKED ISSUE #2003: [FP8] Memory Leak and Fragmentation Investigation
# FP8 Memory Leak Investigation & Reproduction Guide

## 1. Executive Summary

We identified that standard FP8 RL suffers from huge memory growth due to uncleared FP8 scaling factors and workspace buffers. While we can fix the leak by manually clearing caches, this leads to memory fragmentation (high reserved memory). Enabling `expandable_segments:True` solves the fragmentation but introduces a ~33% performance penalty.

## 2. Experiment Results (8-Node Qwen 30B)

We compared 4 configurations to isolate the issue. The **"Memory after refit"** snapshot is the critical metric, representing the "clean slate" memory usage between training steps.

| Configuration | Time/Step | Memory (Allocated) | Memory (Reserved) | Status |
| :--- | :--- | :--- | :--- | :--- |
| **[BF16 (Baseline)](https://wandb.ai/nvidia/grpo-dev-zhiyul/runs/bukjq74s/logs)** | **~95s** | **~0.5 GB** | **~5.2 GB** | ✅ **Gold Standard** |
| **[FP8 (Default)](https://wandb.ai/nvidia/grpo-dev-zhiyul/runs/g1657cyh/logs)** | ~87s | ~8.9 GB | **~26.0 GB** | ❌ **Huge Leak** (Both high) |
| **[FP8 + ClearCaches](https://wandb.ai/nvidia/grpo-dev-zhiyul/runs/ulwnfku5/logs)** | ~91s | ~0.6 GB | **~15.5 GB*** | ⚠️ **Fragmentation** (Allocated low, Reserved high) |
| **[FP8 + Clear + Expandable(de-fregmentation flag)](https://wandb.ai/nvidia/grpo-dev-zhiyul/runs/ncwnsmvm/logs)** | ~127s | ~1.1 GB | **~3.0 GB** | ⚠️ **Fixed Leak, but 33% Slower** |

*\*Note: In the `ClearCaches` run, some ranks showed ~3GB reserved while others showed ~15.5GB, indicating severe non-deterministic memory fragmentation.*

## 3. The Issue: Huge Memory Overhead

In the standard FP8 implementation, we observe massive memory usage compared to BF16. This is driven by three factors:
1.  **FP8 Scaling Factors:** Accumulating `int16` tensors that aren't cleared after use. And why it takes additional **8GB** allocated + **20GB** reserved???
2.  **Workspace Buffers:** Temporary buffers for FP8 operations that persist across steps.
3.  **Fragmentation:** Small, frequent allocations cause the CUDA allocator to reserve large blocks of memory that cannot be efficiently reused, leading to OOMs even when "allocated" memory is low.

### A. FP8 (Default)
*   **Observation:** Both allocated and reserved memory are massive and grow over time.
*   **Root Cause:** FP8 workspaces and scaling factors are not being cleaned up by the framework.

### B. FP8 + `forceClearFP8Caches`
*   **Observation:** `Allocated` memory drops to near baseline (~0.6GB), proving the objects are successfully freed. However, `Reserved` memory remains dangerously high (~15.5GB).
*   **Root Cause:** The CUDA allocator cannot release the fragmented memory blocks back to the pool. The memory is "free" but trapped in unusable chunks.

### C. FP8 + `forceClearFP8Caches` + `expandable_segments:True`
*   **Observation:** Solves the fragmentation. Reserved memory drops to ~3GB (perfect stability).
*   **Trade-off:** **~33% slowdown** (95s → 127s per step).
*   **Hypothesis:** The performance hit comes from the overhead of the allocator managing expandable segments during the high-frequency allocations typical of FP8 training.

## 5. Reproduction Guide

Use the script [`local_qwen_fp8_8node.sh`](https://github.com/NVIDIA-NeMo/RL/blob/zhiyul/fp_memory_cleanup/local_qwen_fp8_8node.sh) to reproduce these states. It is designed to be simple and switchable via environment variables.

* branch: https://github.com/NVIDIA-NeMo/RL/blob/zhiyul/fp_memory_cleanup 
   - close to r0.5.0

```
# interactive mode use ray.sub
CONTAINER=/lustre/fsw/portfolios/coreai/users/zhiyul/enroot-images/nvcr.io#nvidia/nemo-rl:v0.5.0.squashfs\
MOUNTS="$PWD:$PWD" \
sbatch \
    --nodes=8 \
    --account=YOUR_ACCOUNT \
    --job-name=YOUR_JOBNAME \
    --partition=YOUR_PARTITION \
    --time=4:0:0 \
    --gres=gpu:8 \
    ray.sub
```

```
# run the following commands under the container:
bash local_qwen_fp8_8node.sh
```

### Usage Commands

**1. Baseline BF16 (Fast, Low Memory)**
```bash
bash local_qwen_fp8_8node.sh
```

**2. FP8 Default (Fast, High Memory Leak)**
```bash
FP8=true bash local_qwen_fp8_8node.sh
```

**3. FP8 + Fix Leak (Fast, High Fragmentation)**
```bash
FP8=true FORCE_CLEAR_FP8_CACHES=true bash local_qwen_fp8_8node.sh
```

**4. FP8 + Fix Leak + Fix Fragmentation (Slow, Low Memory)**
```bash
FP8=true FORCE_CLEAR_FP8_CACHES=true EXPANDABLE=true bash local_qwen_fp8_8node.sh
```

### How to Verify
Search the logs for the "Refit Complete" memory snapshot. This line appears after model weights are reloaded but before the next training step begins.

```bash
grep "GPU Memory after refit complete" logs_fp_memory/your_log_file
```

## 6. Verification

It would be considerred resolved perfectly if we can reach similar close-to-zero memory allocated (after refit) and reserved with no extra speed costs in fp8.

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

**Add a one line overview of what this PR aims to accomplish.**

# Issues
closes #2003 

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_rl/models/policy/workers/megatron_policy_worker.py [hunk 1]
```diff
@@ -1234,6 +1234,41 @@ def prepare_for_training(self, *args, **kwargs):
...
+                workspace_count += 1
+
+        # 2. Clear global TE workspace
+        try:
+            import transformer_engine.pytorch as te
+
+            if hasattr(te, "module") and hasattr(te.module.base, "clear_workspace"):
+                te.module.base.clear_workspace()
+        except ImportError:
+            pass
```
Comment (ZhiyuLi-Nvidia) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py]: Hi, @ashors1 might be my issue. I can't find this in TE and I don't think this piece code is helpful to the memory saving, maybe we can simply remove it?

File: nemo_rl/models/policy/workers/megatron_policy_worker.py [hunk 2]
```diff
@@ -1234,6 +1234,41 @@ def prepare_for_training(self, *args, **kwargs):
...
+
+        print(
+            f"[_clear_fp8_caches] Cleared {workspace_count} workspace modules on rank {self.rank}"
+        )
+
+        gc.collect()
+        torch.cuda.empty_cache()
```
Comment (ZhiyuLi-Nvidia) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py]: Would it be more efficient to run `gc.collect()` and `torch.cuda.empty_cache()` just once after all references are cleared? Calling them frequently might introduce unnecessary overhead to the training pipeline.

File: nemo_rl/models/policy/workers/megatron_policy_worker.py [hunk 3]
```diff
@@ -1276,6 +1316,58 @@ def offload_after_refit(self):
...
+                        "local_probs",                              # AllGather
+                        "local_map",                                # AllGather
+                    ):
+                        if isinstance(getattr(dispatcher, attr, None), torch.Tensor):
+                            setattr(dispatcher, attr, None)
+            except Exception:
+                pass
+
+            gc.collect()
+            torch.cuda.empty_cache()
```
Comment (ZhiyuLi-Nvidia) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py]: Same above: 
Would it be more efficient to run gc.collect() and torch.cuda.empty_cache() just once after all references are cleared? Calling them frequently might introduce unnecessary overhead to the training pipeline.

**Review by ZhiyuLi-Nvidia — DISMISSED** [review]
Thank you @ashors1 for driving the fix!
Just left some nit.
LGTM!

**Review by ZhiyuLi-Nvidia — DISMISSED** [review]

**Review by ZhiyuLi-Nvidia — DISMISSED** [review]
Thank you @ashors1. LGTM!

**Review by terrykong — DISMISSED** [review]

**Review by terrykong — APPROVED** [review]
